src/vm/methodtable.cpp

```diff
@@ -2703,17 +2722,33 @@ bool MethodTable::ClassifyEightBytesWithNativeLayout(SystemVStructRegisterPassin
             {
                 MethodTable* pFieldMT = ((FieldMarshaler_FixedArray*)pFieldMarshaler)->GetElementMethodTable();
 
-                bool inEmbeddedStructPrev = helperPtr->inEmbeddedStruct;
-                helperPtr->inEmbeddedStruct = true;
-                bool structRet = pFieldMT->ClassifyEightBytesWithNativeLayout(helperPtr, nestingLevel + 1, normalizedFieldOffset, useNativeLayout);
-                helperPtr->inEmbeddedStruct = inEmbeddedStructPrev;
+                unsigned int numElements = ((FieldMarshaler_FixedArray*)pFieldMarshaler)->GetNumElements();
+
+                bool structRet = true;
+
+                for (unsigned int i = 0; i < numElements; i++, normalizedFieldOffset += pFieldMT->GetNativeSize())
+                {
+                    bool inEmbeddedStructPrev = helperPtr->inEmbeddedStruct;
+                    helperPtr->inEmbeddedStruct = true;
+                    structRet = pFieldMT->ClassifyEightBytesWithNativeLayout(
+                        helperPtr,
+                        nestingLevel + 1,
+                        normalizedFieldOffset,
+                        useNativeLayout);
+                    helperPtr->inEmbeddedStruct = inEmbeddedStructPrev;
+                }
 
                 if (!structRet)
                 {
                     // If the nested struct says not to enregister, there's no need to continue analyzing at this level. Just return do not enregister.
                     return false;
                 }
 
+                if (!isFixedBuffer) // We reuse the field marshaler for fixed buffers.
+                {
+                    ((BYTE*&)pFieldMarshaler) += MAXFIELDMARSHALERSIZE;
```

Comment:
This deserves a comment. Whenever pointer arithmetic is being done after a cast to `BYTE*` there needs to be a comment.

Reply:
This snippet it used multiple times. I am really curious what we are offsetting into.

Reply:
The field marshalers are stored sequentially in a single memory block. Each block for a field marshaler is `MAXFIELDMARSHALERSIZE` wide to support any of the derived `FieldMarshaler_*` types in it. This construct here is used basically anywhere field marshalers are used to move along to the next FieldMarshaler in the block.

Reply:
Sure, but that is not obvious. In fact this seems easy to just make a helper function that says `GetNextMarshaller(...)`.

Reply:
Nevermind. I see what you are saying.